server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -49,188 +51,314 @@
+    return numToAssign;
   }
 
-  private CoordinatorStats drop(
-      final Map<String, Integer> loadStatus,
+  private void drop(
+      final DruidCoordinatorRuntimeParams params,
       final DataSegment segment,
-      final DruidCoordinatorRuntimeParams params
+      final CoordinatorStats stats
   )
   {
-    CoordinatorStats stats = new CoordinatorStats();
+    final DruidCluster druidCluster = params.getDruidCluster();
 
-    // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
-      }
+    // This enforces that loading is completed before we attempt to drop stuffs as a safety measure.
+    if (loadingInProgress(druidCluster)) {
+      return;
     }
 
-    // Find all instances of this segment across tiers
-    Map<String, Integer> replicantsByTier = params.getSegmentReplicantLookup().getClusterTiers(segment.getIdentifier());
+    for (final Object2IntMap.Entry<String> entry : currentReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+
+      final NavigableSet<ServerHolder> holders = druidCluster.getHistoricalsByTier(tier);
+
+      final int numDropped;
+      if (holders == null) {
+        log.makeAlert("No holders found for tier[%s]", tier).emit();
+        numDropped = 0;
+      } else {
+        final int currentReplicantsInTier = entry.getIntValue();
+        final int numToDrop = currentReplicantsInTier - targetReplicants.getOrDefault(tier, 0);
+        numDropped = dropForTier(numToDrop, holders, segment);
+      }
 
-    for (Map.Entry<String, Integer> entry : replicantsByTier.entrySet()) {
+      stats.addToTieredStat(DROPPED_COUNT, tier, numDropped);
+    }
+  }
+
+  /**
+   * Returns true if at least one tier in target replica assignment exists in cluster but does not have enough replicas.
+   */
+  private boolean loadingInProgress(final DruidCluster druidCluster)
+  {
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
       final String tier = entry.getKey();
-      int loadedNumReplicantsForTier = entry.getValue();
-      int expectedNumReplicantsForTier = getNumReplicants(tier);
+      // if there are replicants loading in cluster
+      if (druidCluster.hasTier(tier) && entry.getIntValue() > currentReplicants.getOrDefault(tier, 0)) {
+        return true;
+      }
+    }
 
-      stats.addToTieredStat(DROPPED_COUNT, tier, 0);
+    return false;
+  }
 
-      MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
-      if (serverQueue == null) {
-        log.makeAlert("No holders found for tier[%s]", entry.getKey()).emit();
-        continue;
+  private static int dropForTier(
+      final int numToDrop,
+      final NavigableSet<ServerHolder> holdersInTier,
+      final DataSegment segment
+  )
+  {
+    int numDropped = 0;
```

Comment:
(Optional) This can be replaced with java 8 awesomeness

```java
    return StreamSupport
        .stream(Spliterators.spliteratorUnknownSize(holdersInTier.descendingIterator(), Spliterator.ORDERED), false)
        .limit(numToDrop)
        .filter(sh -> sh.isServingSegment(segment))
        .mapToInt(sh -> {
          sh.getPeon().dropSegment(segment, null);
          return 1;
        })
        .sum();
```

Reply:
```java
    return StreamSupport
        .stream(Spliterators.spliteratorUnknownSize(holdersInTier.descendingIterator(), Spliterator.ORDERED), false)
        .filter(sh -> sh.isServingSegment(segment))
        .limit(numToDrop)
        .mapToInt(sh -> {
          sh.getPeon().dropSegment(segment, null);
          return 1;
        })
        .sum();
```

is the correct one I think, but highlights a blind spot in the testing.

Reply:
I think I am going to skip using monad-chaining on this one because we still need to work in the `log.warn()` into it and it does not seem quite worth it anymore 🤷‍♂️

Reply:
I've included additional auxiliary fixture to fail the first implementation.